### ISSUE TITLE
Null<T> lost in type inference

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#5517]
```haxe
class Node<T> {
    public function new(value:T, next:Node<T>) {
    }
}

class Main {
    static function main() {
        var i:Null<Int> = 1;
		new Node(i, new Node(1, null));
    }
}
```

generated AST (note the inner Node is a `Node<Int>`, not `Node<Null<Int>>`):
```
[New:Node<Null<Int>>]
	Node<Null<Int>>
	[Cast:Null<Int>] [Const:Int] 1
	[New:Node<Int>]
		Node<Int>
		[Const:Int] 1
		[Const:Node<Int>] null
}
```

this is problematic for C# target where we have to generate a copy-cast right now

**Comment by nadako** [comment]
The example was taken from Issue3681 which failed to compile for C# with `-D fast-cast` if I fix the redundant `__hx_cast` generation.

**Comment by Simn** [comment]
I don't even know where to start looking here. I'm assuming it's not a regression?

**Comment by Simn** [comment]
Is that actually unexpected? You call `new Node(1, null)`, so T gets bound to Int.

**Comment by nadako** [comment]
Yes, but the problem is that `next` compiler argument receives a different `T` (`Int` instead of `Null<Int>` in this case).

**Comment by Simn** [comment]
Yes, because they unify. It's clearer if you split up the call:

```haxe
var i:Null<Int> = 1;
var next = new Node(1, null);
new Node(i, next);
```

The only problem here is that the second argument is silently cast from `Node<Int>` to `Node<Null<Int>>` because `Null<T>` is a typedef. That's not a type-inference issue though.

You _could_ maybe make the argument that in the merged call, top-down inference should cause the inner type to be inferred to `Null<T>` straight away. I'm not quite sure about that, but it wouldn't solve the problem in the split-up version anyway.

**Comment by nadako** [comment]
> You could maybe make the argument that in the merged call, top-down inference should cause the inner type to be inferred to Null<T> straight away.

Yeah that what I was expecting here originally. The split-up version is more explicit and should cause an error...
